domain/schema/model.go

```diff
@@ -482,6 +482,8 @@ CREATE TABLE storage_instance_pool (
 -- storage_unit_owner is used to indicate when
 -- a unit is the owner of a storage instance.
 -- This is different to a storage attachment.
+-- The owner should be *either* an application
+-- or a unit; not both.
```

Comment:
Are you sure? I can't see anywhere where any owner other than a unit is set.
This includes initial creation and reattachment after being detached.
And all of the tests which assess this attribute expect the value to be a unit.

Reply:
The UML that Andrew (?) produced has applications and units as owners. It makes a shared/non-shared distinction. Note that this is separate from attachment.

It is also explicitly stated [here](https://github.com/juju/juju/blob/19c22e5b0c62aaf557a537b0b63d76adcd03a153/state/storage.go#L41).

Reply:
Regarding volume-backed filesystems, the question is whether those can exist separately from a storage instance. If they can, we need the join table.

If they can't, I'd like to see how this approach works out - we can always add the join table later.

The reason is that although the application logic would prevent it, with a join table there would be nothing in the schema design that forbids a storage instance *without* both entities from a volume-backed filesystem pair being linked to it.

Reply:
> The UML that Andrew (?) produced has applications and units as owners. It makes a shared/non-shared distinction. Note that this is separate from attachment.
> 
> It is also explicitly stated [here](https://github.com/juju/juju/blob/19c22e5b0c62aaf557a537b0b63d76adcd03a153/state/storage.go#L41).

Ah right, that's for shared storage which was loosely spec'ed at the time but never implemented. That's why I couldn't find any uses of it. IMO we shouldn't introduce artefacts related to something we never implemented, but so long as the design doesn't preclude being extended later as needed, that would be sufficient.

Reply:
> Regarding volume-backed filesystems, the question is whether those can exist separately from a storage instance. If they can, we need the join table.
> 

Currently, filesystem and volume entities are always tied to a storage instance - they are created as an artefact of the logical storage provisioning operation. It's the storage instance that is tied to an owner. So filesystems and volumes are not created standalone.

> If they can't, I'd like to see how this approach works out - we can always add the join table later.
> 
> The reason is that although the application logic would prevent it, with a join table there would be nothing in the schema design that forbids a storage instance _without_ both entities from a volume-backed filesystem pair being linked to it.

That's still the case now though right? Anyway, no strong view on either approach.

Reply:
Right, so I've reverted the application ownership addition. This should be good to land and add info/params to if you're OK with it.